### ISSUE TITLE
tmt should attempt to reinstall test requires before a test run

### ISSUE THREAD

**Issue by kkaarreell** [teemtee/tmt#1862]
`tmt` installs test dependencies during the `prepare` step. However, if packages are modified by a test it may be missing later. It would be nice if `tmt` would check and eventually attempt to install missing test requires before each test. It could be conditioned by some cmdline option or some `prepare` step option. Definitely doesn't have to be enabled by default.

**Comment by psss** [comment]
Looks like a reasonable feature if not enabled by default as it could significantly slow down the test execution.

**Comment by jscotka** [comment]
Hi,
theoretically I can imagine some clever optimization. check package list changes. like using dnf transactions, logs or using some `inotify` on rpm database. and checking just reinstall if any required package changed. It is more less just optimization, but should be then more less fast if nothing changes between  tests.

**Comment by adiosnb** [comment]
What do you think about a `check-dependencies`: `bool` parameter (`false` by default) for the execution step, which would enable the discussed behavior?

It indeed can check a difference of `rpm -qa` before the test execution, and install dependent packages, if necessary.

**Comment by happz** [comment]
> What do you think about a `check-dependencies`: `bool` parameter (`false` by default) for the execution step, which would enable the discussed behavior?
> 
> It indeed can check a difference of `rpm -qa` before the test execution, and install dependent packages, if necessary.

Sounds like (yet another) check from https://github.com/teemtee/tmt/pull/216

**Comment by kkaarreell** [comment]
> Sounds like (yet another) check from #216

I would disagree. This RFE is supposed to address `tmt` design decision that all test dependencies will be installed before tests are executed. This is frankly both naive and limiting assumption.
A check inspecting whether specific NVRs are installed, that would be a material for #216 but not this RFE itself (IMHO).

**Comment by adiosnb** [comment]
> > What do you think about a `check-dependencies`: `bool` parameter (`false` by default) for the execution step, which would enable the discussed behavior?
> > It indeed can check a difference of `rpm -qa` before the test execution, and install dependent packages, if necessary.
> 
> Sounds like (yet another) check from #216

This is more like a check + command. Maybe `force-dependencies`:`boo` would be better.

**Comment by happz** [comment]
> > Sounds like (yet another) check from #216
> 
> I would disagree. This RFE is supposed to address `tmt` design decision that all test dependencies will be installed before tests are executed. This is frankly both naive and limiting assumption. A check inspecting whether specific NVRs are installed, that would be a material for #216 but not this RFE itself (IMHO).

Ok, not a check then, but something more fundamental. Yet, "checking" things here and there is mentioned a lot in this issue, that's probably what put me on this track.

Anyway, so, are we talking more like the restraint approach, installing test requirements/packages under the test before each test? Some kind "save rpm list & restore" action?

**Comment by lukaszachy** [comment]
> Anyway, so, are we talking more like the restraint approach, installing test requirements/packages under the test before each test? Some kind "save rpm list & restore" action?

I think it is more naive way - run prepare for require & recommend again. 

Best way would be if test does proper thing (e.g https://github.com/beakerlib/rpm/tree/master/snapshot/) and don't modify the packages.

**Comment by kkaarreell** [comment]
It's not just about modyfying, tests may be running on conflicting packages. There are multiple reasons.

**Comment by kkaarreell** [comment]
This is unfortunate. It means we won't be able to optimize test plan execution but rather maintain multiple plans and utilize more test systems.

> Please follow similar discussions:
>     * `isolate`: https://tmt.readthedocs.io/en/latest/spec/plans.html#isolate
>     * `destructive`: [Support tagging a test as "destructive" #524](https://github.com/teemtee/tmt/issues/524)

Those doesn't solve my problem. Anyway, thanks for considering this.
